kyuubi-hive-jdbc/src/main/java/org/apache/kyuubi/jdbc/hive/auth/CachingKerberosAuthentication.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kyuubi.jdbc.hive.auth;
+
+import static java.util.Objects.requireNonNull;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.security.auth.Subject;
+import javax.security.auth.kerberos.KerberosTicket;
+
+public class CachingKerberosAuthentication {
+  private final KerberosAuthentication kerberosAuthentication;
+
+  @GuardedBy("this")
+  private Subject subject;
+
+  @GuardedBy("this")
+  private long nextRefreshTime;
+
+  public CachingKerberosAuthentication(KerberosAuthentication kerberosAuthentication) {
+    this.kerberosAuthentication =
+        requireNonNull(kerberosAuthentication, "kerberosAuthentication is null");
+  }
+
+  public synchronized Subject getSubject() {
+    if (subject == null || ticketNeedsRefresh()) {
+      subject =
+          requireNonNull(
+              kerberosAuthentication.getSubject(), "kerberosAuthentication.getSubject() is null");
+      KerberosTicket tgtTicket = KerberosUtils.getTgt(subject);
+      nextRefreshTime = KerberosUtils.getTgtRefreshTime(tgtTicket);
+    }
+    return subject;
+  }
+
+  public synchronized void reauthenticateIfSoonWillBeExpired() {
```

Comment:
This method is not used. Will we spawn a thread like `UserGroupInformation#spawnAutoRenewalThreadForUserCreds`  to refresh TgtCache ?

Reply:
It's a good point, since `CachingKerberosAuthentication#getSubject` is called on creating each connection, then in the current implementation, it would be a problem if the user creates a connection and keeps it for a long time, especially when the user uses the connection pool.
Considering this PR is large enough, I prefer to implement the `spawnAutoRenewalThreadForUserCreds` in the following PR.